httpbp/response.go

```diff
@@ -107,6 +107,84 @@ type ErrorResponse interface {
 	Err() error
 }
 
+// BaseResponse can be embedded into other response structs to allow them to
+// implement the go-kit Headerer and StatusCoder interfaces as well as the
+// baseplate.go ResponseCookies and ErrorResponse interfaces.
+//
+// BaseResponse must be initalized using NewBaseResponse before use, if it is
+// not, some methods will panic.
+//
+// 		type Response struct {
+//			httpbp.BaseResponse
+// 		}
+//
+//		func NewResponse() *Response {
+//			return &Response{
+//				BaseResponse: httpbp.NewBaseResponse(),
+//			}
+//		}
+type BaseResponse struct {
+	code    int
+	headers http.Header
+	cookies []*http.Cookie
+	err     error
+}
+
+// NewBaseResponse returns an initialized BaseResponse.
+//
+// Intended to be used by the constructor methods for Response structs that
+// embed BaseResponse.
+func NewBaseResponse() BaseResponse {
+	return BaseResponse{headers: make(http.Header)}
+}
+
+// SetCode sets the status code for this response.
+func (r *BaseResponse) SetCode(code int) {
+	r.code = code
+}
+
+// StatusCode returns the current status code set for this response.
+func (r BaseResponse) StatusCode() int {
+	return r.code
+}
+
+// Headers returns the http.Header collection of headers to set on the response.
+func (r BaseResponse) Headers() http.Header {
+	return r.headers
+}
+
+// SetCookie adds a cookie to set on the response.
+func (r *BaseResponse) SetCookie(cookie *http.Cookie) {
+	r.cookies = append(r.cookies, cookie)
+}
+
+// Cookies returns the current list of cookies to set on the response.
+func (r BaseResponse) Cookies() []*http.Cookie {
+	return r.cookies
```

Comment:
In go, returning a slice field in a function makes the ownership muddy. Say we have 3 items in the slice, the caller can actually change the 3 items inside the slice to other values, but they cannot add new items nor remove the size of the slice.

As a result, if the intention is to give read-only access, making a copy is recommended:

```go
cookies := make([]*http.Cookie, len(r.cookies))
copy(cookies, r.cookies)
return cookies
```

Reply:
note that if after `Cookies()` is called we no longer care about the underlying cookies (e.g. the object should be gc'd after that) then this is fine.

Reply:
In the way that `baseplate.go` uses `Cookies()`, we don't care about it _but_ I think it's still a good idea for cases where we want to use it outside of just the `EncodeResponse` methods in baseplate.